pkg/config/getters.go

```diff
@@ -161,3 +164,144 @@ func GetStoragePath() string {
 func PreferredAddress() string {
 	return os.Getenv("BACALHAU_PREFERRED_ADDRESS")
 }
+
+func GetStringMapString(key string) map[string]string {
+	return viper.GetStringMapString(key)
+}
+
+func Get[T any](key string) (T, error) {
+	raw := viper.Get(key)
+	if raw == nil {
+		return zeroValue[T](), fmt.Errorf("value not found for %s", key)
+	}
+
+	var val T
+	val, ok := raw.(T)
+	if !ok {
+		err := ForKey(key, &val)
+		if err != nil {
+			return zeroValue[T](), fmt.Errorf("value not of expected type, got: %T", raw)
+		}
+	}
+
+	return val, nil
+}
+
+func zeroValue[T any]() T {
+	var zero T
+	return zero
+}
+
+// ForKey unmarshals configuration values associated with a given key into the provided cfg structure.
+// It uses unmarshalCompositeKey internally to handle composite keys, ensuring values spread across
+// nested sub-keys are correctly populated into the cfg structure.
+//
+// Parameters:
+//   - key: The configuration key to retrieve values for.
+//   - cfg: The structure into which the configuration values will be unmarshaled.
+//
+// Returns:
+//   - An error if any occurred during unmarshaling; otherwise, nil.
+func ForKey(key string, cfg interface{}) error {
+	return unmarshalCompositeKey(key, cfg)
+}
+
+// unmarshalCompositeKey takes a key and an output structure to unmarshal into. It gets the
+// composite value associated with the given key and decodes it into the provided output structure.
+// It's especially useful when the desired value is not directly associated with the key, but
+// instead is spread across various nested sub-keys within the configuration.
+func unmarshalCompositeKey(key string, output interface{}) error {
+	compositeValue, isNested, err := getCompositeValue(key)
+	if err != nil {
+		return err
+	}
+	decoderConfig := &mapstructure.DecoderConfig{
+		DecodeHook: mapstructure.TextUnmarshallerHookFunc(),
+		Result:     output,
+		TagName:    "mapstructure", // This is the default struct tag name used by Viper.
+	}
+
+	decoder, err := mapstructure.NewDecoder(decoderConfig)
+	if err != nil {
+		return err
+	}
+
+	if isNested {
+		val, ok := compositeValue[key]
+		if !ok {
+			// NB(forrest): this case should never happen as we ensure all configuration values
+			// have a corresponding key via code gen. If this does occur it represents an error we need to debug.
+			return fmt.Errorf("CRITICAL ERROR: invalid configuration detected for key: %s. Config value not found", key)
```

Comment:
Where this eventually unrolls to ForKey() the error we set here is ignored and "value not of expected type, got: %T" return instead.  Should we log the error as well if it is important?

Reply:
Great catch. I've adjusted the error message there to include the original error. I will also add a log line here.